### ISSUE TITLE
Implement an admin interface where options/candidates to be voted for can be added

### ISSUE THREAD

**Issue by anaoprea** [anaoprea/iDibo#11]
First of all, think about a way you could allow different user to have different authorization levels.

For example, if the students have to vote for what they would like to have for lunch more often (fruits or sweets as desert), then a student who can register and vote is a "regular" user and the secretary from the administration office, who can add in the system the poll and the options, is an "admin" user. Read a bit about Role Based Access Control [1]
1. Implement a simple RBAC structure: for example, you could have a "Role" in the "User" table created before.
2. The admin users, are able to create campaigns and add options (for example, create the campaign "Lunch dessert" that has a description for what it means, a period when it is active and the options fruits and sweets).
3. The regular users cannot create campaigns and, once they logged in, can see the active campaigns and vote. The logic for the voting will be added in a future issue.
4. Add tests for validating an admin user can create a campaign and a regular user cannot, but can see an active campaign.

[1] http://en.wikipedia.org/wiki/Role-based_access_control

**Comment by OluyemisiAkin** [comment]
For the RBAC, i'm looking at 2 options: 1) adding a new field in the registration form for the user to select his/her role. 2) Having registration options, those registering as users or dose registering as admin and then have the privileged to add topics like the way we have student and staff account is school portals

**Comment by OluyemisiAkin** [comment]
What do you say to the proposed methods?

**Comment by OluyemisiAkin** [comment]
Here is a full picture of what I'm considering:
An extra field(a drop down menu) is added to the registration fields where user can choose one of two roles, admin user/regular user.
-Regular users can just sign up easily
-But users that choose admin user needs to be verified via email(i'm thinking separating this might be a task though), or lets say a new field shows up which is admin code or something like a staff id that is unique to the user(just a form of identification).
Based on  this i can now mange the pages for each user based on his group or role.

**Comment by anaoprea** [comment]
I think that the solution is good, please proceed to implementing it.

Please take into account the fact that it might be useful to also validate the normal user registration against the voter's ID (maybe respect a pattern).

**Comment by anaoprea** [comment]
Regarding "After creating the roles, i cant figure out how to link it to determining which confirm page to lead the user to on register or log in. as i've already set login_next to cofirm.html permanently.": the confirmation page should always be the same, something like "Congratulations for registering as a <role_here>".

What will be different is the contents of the profile after logging in: the user will have candidates to vote for, the admin will have campaigns to administer - in both cases, this is dynamic information that you are retrieving from the database => you have to first think about the database schema design.

Hint: there should probably be a campaigns table, please think about what fields it should have to represent the various options.

**Comment by OluyemisiAkin** [comment]
Yes, I actually thought about that and I'm working on it right from the contoller. The other issue is, what is loaded for either user is suppose to be kinds if dynamic, something like the list of candidates to vote for is updated from a table or so, and for admin a list of campaigns he has administered and option to add more. 
So how do I work with that, do I need to create new tables for these data?

Best Regards,
'Yemisi

**Comment by anaoprea** [comment]
Let's imagine a simplified scenario, based on majority for winning:
- A user can only vote for campaigns that are added after she/he has registered.
- A user can only vote once and cannot change the vote after (stateless system).
- All the users can vote for all the active campaigns (further refinement based on group rights should be added later).
- An admin creates a campaign that everyone is allowed to vote on: the information about the campaign is stored in a **campaigns** table. Furthermore, for each user existent in the system, a new row is added to the **participation** table, with the _campaign_ and _user_ IDs and a default value that the user has not yet voted.
- A user sees the existent campaign and votes: there is an **options** table, that has a foreign key to the **campaign** table and holds a _counter_ that will be increased every time a user chooses the option. Furthermore, to make sure that everyone can vote only once, the _has_voted_ flag is set to _false_ in the **participation** table.
  When the _campaign_ has ended, you will go to the **options** table, search for all the options based on a _campaign_id_, compute who the winner is based on which option has the highest counter. This information can be displayed in both user and admin profile.

As a later feature, different statistics could be displayed in the admin profile: what percentage of users (from those who were allowed to) voted, what percentage each option had, etc.

**Comment by anaoprea** [comment]
![litg_db](https://cloud.githubusercontent.com/assets/8439892/5942465/f7751ed4-a6fe-11e4-9c7f-d18ad9cc56da.png)

**Comment by anaoprea** [comment]
Please let me know your opinion on the DB design above: I would like to know what you think should be different and if it is clear how you would implement the operations described in the simplified scenario.

**Comment by OluyemisiAkin** [comment]
I have nothing to add for now, i think its ok and illustrates properly what we want to achieve.
